src/client/canvas/renderer.ts

```diff
@@ -43,40 +43,53 @@ export class CanvasRenderer {
     this.context.rotate(-camera.rotate)
     this.context.translate(translateDash.x, translateDash.y)
 
-    const renderObjects = (objects: Scene) => {
-      for (const obj of objects) {
-        if (Array.isArray(obj)) {
-          renderObjects(obj)
-          continue
-        }
-        if (obj.geometry instanceof ArrowGeometry) {
-          this.renderArrow(obj)
-          continue
-        }
-        if (obj.geometry instanceof CircleGeometry) {
-          this.renderCircle(obj)
-          continue
-        }
-        if (obj.geometry instanceof LineGeometry) {
-          this.renderLine(obj)
-          continue
-        }
-        if (obj.geometry instanceof MarkerGeometry) {
-          this.renderMarker(obj)
-          continue
-        }
-        if (obj.geometry instanceof PolygonGeometry) {
-          this.renderPolygon(obj)
-          continue
-        }
-        if (obj.geometry instanceof TextGeometry) {
-          this.renderText(obj, camera)
-          continue
-        }
+    this.renderObjects(scene.children, camera)
+
+    this.context.restore()
+  }
+
+  private renderObjects(objects: Object2d[], camera: Transform) {
+    for (const obj of objects) {
+      this.context.save()
+      const translateDash = Vector2.from(obj.transform.translate).applyTransform({
+        rotate: -obj.transform.rotate,
+        scale: {
+          x: 1 / obj.transform.scale.x,
+          y: 1 / obj.transform.scale.y,
+        },
+        translate: {
+          x: 0,
+          y: 0,
+        },
+      })
+      this.context.scale(obj.transform.scale.x, obj.transform.scale.y)
+      this.context.rotate(-obj.transform.rotate)
+      this.context.translate(translateDash.x, translateDash.y)
```

Comment:
Could we factor this above code out into a `applyTransform(context, transform)` function?